### DISCUSSION TITLE
RS485 Connection when ESP is powered from the Battery?

### DISCUSSION THREAD

**Discussion by kinsi55** [syssi/esphome-jk-bms#178] [Q&A] [answered]
I am wondering how the wiring is supposed to look when the ESP is powered from the battery itself (Using a buck converter). Would you still want to connect ground together? (And thus, connect the P- to the GND pin of the RS485 connector).

**Comment by syssi** [comment]
Sorry, I cannot support here. I'm a software developer. ;-) I hope someone else could step in here.

**Comment by kinsi55** [comment, marked as answer]
Ok I found the answer to this question myself: When you power the ESP from the Battery output (After the BMS), absolutely dont connect the GND, otherwise the coprocessor for handling RS485 (And I guess the LCD too seeing as I cannot start / turn off the BMS with the included button any more) will die a quick death (Sizable spark when connecting, for a reason that I dont fully understand).

Luckily it can still be started with the old "jumpstart" method and seems to otherwise function fine. oh well, I guess I wanted to use BLE anyways (I didnt)

**Edit:** I'm starting to even doubt that this would work without connecting GND as the tx pin would of course get pulled to the "Ground" in communication. I guess one would need to use optocouplers or an isolated DC:DC converter

**Reply by syssi** [reply]
Could you draw a schematics to explain what did you try in detail? I would like to be able to write a warning to the documentation.

**Reply by syssi** [reply]
@meccip Could you share your knowledge here too? It sounds like both of you have tried the same without success.

**Reply by syssi** [reply]
See https://github.com/syssi/esphome-jk-bms/discussions/175#discussioncomment-3698318

**Reply by syssi** [reply]
@logmod Please be careful while playing around with `VBAT` to power a daughter board. I'm aware of these two enthusiasts now who have destroyed the serial communication capability of their BMS.

Another discussion for reference: https://github.com/jblance/mpp-solar/issues/112#issuecomment-1250978348

**Reply by meccip** [reply]
The buck converter makes a spark because of the coil ferrite. The shock at start up is fatal for the unprotected GPS port. Looks like that RS485 adapter that sells with bms have galvanic protection inside and is powered directly to V+ bms. The safest method is to power esp separately and connect gnd together. **First power esp and then connect plug to bms !**

**Reply by syssi** [reply]
May be some macro photos of the Jikong RS485 dongle are helpful: https://github.com/syssi/esphome-jk-bms/tree/main/images/rs485-adapter

**Reply by kinsi55** [reply]
The only thing that I had done was essentially connect the RS485 GND to the P- (negative output) of the BMS. I was not (planning on) using VBAT

Thinking about it, it makes sense that this should not be done, but not doing that would probably not have saved me anyways due to the whole TX to P- story

**Reply by logmod** [reply]
I heard a lot of stories about the RS485 module working fine at first and then dying after some days/weeks. So it does not seem to be fully immune against grounding issues or inrush current as well. Looking at what @kinsi55 wrote it seems that the ESP was powered on, otherwise he would not have observed a spark when connecting. For me the spark points more into the direction of a grounding issue. As indicated by @syssi there is a 122U31 chip on the BMS board so there is galvanic isolation for the GPS port. If you power your device directly from the battery then your TTL GND may be at a totally different level. It may help to measure the voltage difference between the GPS port (GND and VCC) and the battery connectors. It would be interesting to know if VCC is powered directly from the battery level.

**Reply by kinsi55** [reply]
The ESP was not part of the equation at that point in time, I merely prepared all the wiring for it and thus made the RS485-GND <> P- connection, that's where I got the spark and the BMS turned off the output, alongside an alert (CPUEXT anomaly)

After rebooting (via Jumpstart), this error is now gone, the LCD port button does still not function tho. I might try and check if I get any communication from the RS485 port with a laptop and RS232 adapter

**Reply by meccip** [reply]
@kinsi55 
I think your problem is different from mine. You connected the bms GND to the battery GND, which resulted in a bypass of the bms. The BMS controls the GND of the battery through mosfetts and you put it on directly.
I used GND and V+ bms to directly power esp. But the buck converter shock destroyed the gps port

**Reply by kinsi55** [reply]
Yeah, FWIW its probably worth mentioning that the Discharge was enabled at that point in time, so it was not down to a big current running though the RS485 GND (or well, shouldn't have been, clearly there was a big current)

**Reply by kinsi55** [reply]
Good news however, I have chatted with the seller (Hankzor) and it seems like they are willing to provide a replacement. I will be keeping this unit for now as other than the afformentioned problem, it seems to still work perfectly fine. They would offer a replacement for the entire warranty period (1 year) so if any issues should arise at a later point (Or somebody smarter than me figures out this problem) I still have that option.

**Reply by kinsi55** [reply]
To add a bit more to the postmortem: To make absolutely sure that I'm not a complete idiot, nor that I received some weird unit, I just measured the other side of the connector and it indeed is VBAT - However, unlike before, there is no more continuity between GND and B-. I wonder if the GND of that connector is somehow isolated in some special way and I managed to fry just that. The `CPUEXT anomaly` alert that I had after the incident was gone after restarting the BMS. I wont bother testing that theory for now tho as I'd like to avoid sending the BMS to heaven all the way.

**Reply by syssi** [reply]
@kinsi55 Could you contact me via mail (basti@linkt.de)? I'm still looking for some more details.

**Reply by kinsi55** [reply]
@syssi Do you happen to have Twitter or Discord? I'd prefer that over mail

**Reply by syssi** [reply]
@kinsi55 I'm available via the Home Assistant and ESPHome discord (Nickname: syssi, \#5175) or twitter (\@syssi84).

**Comment by logmod** [comment]
@kinsi55 Previously you wrote that you connected the buck converter between B+ and P-. Later you write that you checked the connection between GND and B- and found that they were linked.  Did you also measure a link between P- and GND? 

The fact that you now see a different connectivity could just mean this is the result of the destroyed chip. BTW, did you connect your P- to the earth ground of your house (which you should absolutely do)? 

JK chose to expose the full VBAT on the GPS plug, probably because they wanted more flexibility to use the power as needed in different external modules. Their RS485 module comes with an isolated DC converter and an isolated RS485 driver. If you want to power your ESP though a battery powered DC converter you probably should use the same approach and convert your DC from this port.

**Reply by kinsi55** [reply]
There was no buck converter, nor ESP. I was preparing the wiring for those and in that process connected the RS485-GND to P- (As that is what the ESP's GND would eventually become). Retrospectively, this of course was inevitably going to cause issues whenever I turn off Discharge - At the point in time when I did connect it tho, it was enabled, which is what confuses me as to how it would result in my outcome.

You read correctly that I only measured RS485-GND to B- as that measurement was easier than measuring it to P-

My P- is not connected to earth, and I've not heard before that it should be? I dont think I ever saw any mention of this, do you have any reference?

**Reply by logmod** [reply]
I think that you put the wrong caption for the discussion then. There is no buck-converter and no ESP, you just connected the GPS GND and P- (for whatever reason) and ruined your communication port  :-)  

Grounding high voltage stationary battery systems is a common practice, you can read it in many national codes and also in the installation manuals for commercial battery systems. NEC states that bonding between DC and ground is optional (but recommended) for DC systems that operate at less than 50 volts while systems that operate at 50V or more are required to be bonded to ground. So you could argue that it is not mandatory for 48V systems, but in practice there are good reasons to do it. Some countries say that the border is at 60V, but as grounding is a safety feature my position is to do it also for 48V unless there is a clear reason against it.

**Reply by kinsi55** [reply]
Well the only reason I did make that connection is that, when I was going to do what the title suggests, thats what would have been the setup, but of course I did not get that far

Regarding the grounding bit, I'll look into that, thanks for the headsup

**Reply by logmod** [reply]
Please bear in mind that I do not have the BMS or the RS485 module in my hands yet so I am only making guesses about the use of the digital isolator chips.  If you use the GND level available on the GPS port for the ESP you should be fine. No need to connect anything else on the negative side of your buck-converter.

**Reply by kinsi55** [reply]
Yeah had I done that, I would probably have been fine, but in that case in rare circumstances (Say, the battery goes completely flat and the BMS turns off) the ESP could, unless extra measures are taken (And those actually work), kill the battery as it would not be turned off

**Reply by logmod** [reply]
Not sure to understand the risk of this scenario. If the BMS turns off the inverter due to low voltage it will do this at a threshold defined by you. Let's assume you defined it to leave 5% SOC. Then you have around 500Wh left assuming a typical battery size.  Your ESP will run for months on this.

**Reply by kinsi55** [reply]
The BMS has two "Low voltages" - One is where it turns off discharging, and another, lower point where it turns itself off entirely (Usually, the absolute minimum of the batteries) and you need to start it again manually.

How likely is this to happen in a stationary setting with a reasonable minimum voltage? Not very, regardless I still think its not exactly good practice to power something directly off the batteries, not just because of low voltage protection but also because of all the other faults that can occur.

**Reply by logmod** [reply]
If you use the GPS port VCC pin to power your device I would strongly assume that a BMS shutdown cuts that voltage, too.

**Reply by kinsi55** [reply]
Thats a good point, I'll have to test that theory if / when I should ever get a replacement unit :D

**Comment by gepetto666** [comment]
same question. have powered the d1 with dc dc converter. B+ and P- , but if turn off from bms discharge , the d1 still remain powered whit ground from rs485 connector.
has anybody try to use optocoupler ?

**Reply by meccip** [reply]
I use Esp01 and is connected to jk power supply. (no extra wires for power)

I use the LCD port to take 12v supply + and GND (bms comes with a start switch connected to LCD port, you need to make modification).
You still need to connect GPS port GND to esp01 GND because is not working communication.
I use +12v and GND from LCD port and RX TX GND from GPS port. For 12v to 3.3v I use a stepdown converter AMS1117.

**Reply by gepetto666** [reply]
> I use Esp01 and is connected to jk power supply. (no extra wires for power)
> 
> I use the LCD port to take 12v supply + and GND (bms comes with a start switch connected to LCD port, you need to make modification). You still need to connect GPS port GND to esp01 GND because is not working communication. I use +12v and GND from LCD port and RX TX GND from GPS port. For 12v to 3.3v I use a stepdown converter AMS1117.

I'll try , is a ideea, thnks

**Reply by meccip** [reply]
Dont use dc stepdown converter on 40v+ GPS port because you will create a spark at connection and burn RS485 cip !! (I did that and i burn communication).

**Reply by lucize** [reply]
@meccip I think I have this issue, somehow I think I shorted the wires and now ttl is not working anymore, what is the chip that should be verified?
I didn't opened the case, at the moment. is the pi122U31 ?
now, I measure -1.5v on the ttl ports

**Reply by meccip** [reply]
Romania?

**Reply by lucize** [reply]
yes, as written on the profile..

**Reply by meccip** [reply]
ce folosesti ? cablu ttl sau esp rx tx ? ce sursa de alimentare folosesti, externa sau te-ai legat de + si gnd de la bms ?

**Reply by lucize** [reply]
m-am legat direct la esp32, esp-ul a fost alimentat din usb separat, 2 variante:
1-sursa externa
2-stepdown de pe baterie ~13v (nu din bms) cu un modul fast-charge QC3.0 OKY3399-9A 12-24V

s-ar putea ca atunci cand am trecut pe varianta 2 sa se fi stricat sau am făcut scurt intre pinii de la esp32, cert este ca nu mai merge portul, varianta 2 a mers anterior cu reconectări destul de dese

**Reply by lucize** [reply]
senzorul de temperatura arata date eronate sau N/A cand este conectat portul serial, butonul de pornire pare sa nu mai funcționeze, dar se pornește automat

**Reply by meccip** [reply]
Opreste bms-ul si scoate mufa cu fire rosii care are + . Lasa 1 minut si cupleaza inapoi si porneste bms

**Reply by meccip** [reply]
Prima oara eu am alimentat direct din mufa rs485 unde ai tensiunea bateriei (in cazul meu 50v), am facut un stepdown si la conectare a facut un soc si a intrat bms-ul in protectie si nu a mai mers ttl. Am cumparat alt bms.
A doua oara m-am legat cu alimentarea de portul unde bagi butonul de pornire. Acolo aveam 12v si a functionat pentru o perioada de timp cand a inceput sa se inchida bms-ul fara nicio eroare. Trebuia sa ii scot mufa cu + ca sa isi revina.

Apoi am renuntat la alimentarea din bateria si am pus un transformator de telefon 5v bagat in priza la 220v, si cu GND comun cu iesirea RS485. Au trecut mai bine de 5 luni si nu s-a mai oprit bms-ul. Concluzia este ca iesirea TXRX din bms este foarte sensibila, din aceasta cauza JK vinde modulul de conectare la portul ttl si il transforma in RS485 pentru a avea protectie la tensiune venita din exterior.

**Reply by lucize** [reply]
ok, sa inteleg ca nu ai investigat chip-urile de pe placa, din păcate tot eronat îmi arata temperatura si la ora asta nu m-as pune sa reconectez esp-ul. deconectasem anterior cateva secunde, aveam ceva speranțe ca își revine daca tin mai mult deconectat.
mersi de input! as mai avea si portul de CAN sa merg pe RS485 nativ daca nu reușesc cu ttl-ul

**Reply by meccip** [reply]
Unde arata temperatura eronat, in app-ul bms sau in datele din esp ?
Am desfacut bms-ul dar nu am descoperit nimic.

**Reply by lucize** [reply]
prin Bluetooth, ttl nu mai citește nimic..

**Reply by meccip** [reply]
si in app-ul jkbms pe telefon, temperaturile arata bine ?

**Reply by lucize** [reply]
nu..
tot ce arata in app arata si in Bluetooth, dar acum am măsurat pinii si nu-mi mai dau -1.5v, incerc mâine(azi pe lumina) sa conectez esp-ul

**Comment by paulsteigel** [comment]
Regarding powering of the ESP module, why dont we use a buck converter DC-3.3V (VCC) from the B+ and B- of the battery pack to avoid turning off of the JK-BMS. Put the converter and the ESP module in a safe box outside nearby the JK-BMS.  Also use a relay to trigger the BMS starting up when needed using VCC wired to B- and GND to P-.
I am using ESP8266-01S (with only 4 GPIO available)

**Reply by meccip** [reply]
It is a strange thing powering esp from battery. First time i use buck converter from battery and i burn bms communication. Second time i use issolated source powered from lcd 12v output and my bms shutdown from time to time without reson. In this moment i use 220v/5v transformer with common ground and past more than 8 month without any issues.
